### PR TITLE
[KEYCLOAK-14046] Include groups in user creation via Admin Console

### PR DESCRIPTION
**Pull request by clementcur**
Since the POST /users API now supports providing groups membership, here is the UI
part!

- Added a field in the user creation UI to specify groups the newly created user
will be joining
- Added associated messages in english language

create-user-w-groups_00
create-user-w-groups_01
create-user-w-groups_02
create-user-w-groups_03
create-user-w-groups_04

### CODE REVIEW

**Review by stianst — CHANGES_REQUESTED** [review]
Thanks, but I'm not in favour of adding this. Groups are not used by everyone, and does not belong on the basic create user form, but rather as a separate step in my opinion.

I suggest discussing this on the developer mailing list, and also commenting on the designs for the new admin console to raise this as a use-case, which can be considered in the new designs.

**Comment by stianst** [comment]
Just found KEYCLOAK-11621 and understand now better why this is relevant.

**Comment by clementcur** [comment]
ok, no worries - FYI @stianst, I just posted a thread, see https://groups.google.com/d/topic/keycloak-dev/mOxO_ET4g2w/discussion

**Comment by stianst** [comment]
@ssilvert there's times when an admin is only permitted to create a user if the user is created in a specific group. As the current console doesn't allow you to do it I think this is probably the simplest way to add it to the current admin console. As we discussed in the new console this should be done with a wizzard right, so the slightly strange behaviour of having a groups field on a new user would be redesigned there anyways.

**Comment by ssilvert** [comment]
@clementcur This PR has gotten rather stale.  Are you still interested in seeing this get merged?

From the comments I'm not sure if @stianst approves of doing this now or if he only wants to fix it in the new admin console.

**Comment by clementcur** [comment]
Hi @ssilvert & @stianst, I definitely want to see my proposal merged!

Since the new admin console can take some time to replace the current one with feature parity, I'd prefer to go ahead with this solution for the time being. 

What modifications are needed to see this merged?

**Comment by ssilvert** [comment]
@clementcur I apologize.  I just read @stianst comment again and it looks like he is in favor of this.  I'm looking at it now, but can you go ahead and fix the conflicts?

**Comment by ssilvert** [comment]
@clementcur This will need an integration test.  Nothing fancy.  It just needs to be a good regression test to make sure it always works.  You can add it to this part of the testsuite: https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/users/UserAttributesTest.java

**Comment by clementcur** [comment]
@ssilvert I've added some UI tests, let me know if it's what you had in mind. The build is failing for some reasons unrelated to my changes... It builds & tests OK on my local env. though.
@stianst is there any change you'd like to see in this review?

**Comment by clementcur** [comment]
ping @ssilvert

**Comment by clementcur** [comment]
@stianst, seems like you left some requested changes open.

**Review by stianst — COMMENTED** [review]
I'm okay with this. @ssilvert @vmuzikar can you review this one please?

**Review by ssilvert — CHANGES_REQUESTED** [review]
@clementcur The group selector can't find nested groups.  It only finds groups at the root.

**Comment by clementcur** [comment]
Thanks for your feedback @ssilvert, I've pushed a new commit that flattens nested groups as an array. Let me know if anything else is missing.

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/groups/AbstractGroupTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import org.keycloak.testsuite.console.page.groups.Groups;
+
+import static org.keycloak.testsuite.util.URLAssert.assertCurrentUrlEquals;
+import static org.keycloak.testsuite.util.URLAssert.assertCurrentUrlStartsWith;
+
+/**
+ *
+ * @author clementcur
+ */
+public abstract class AbstractGroupTest extends AbstractConsoleTest {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/groups/AbstractGroupTest.java:19]: Even though this is out of scope of this PR, thanks for adding it. It's a shame we didn't have UI tests for groups. :(

File: testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/users/UserAttributesForm.java [hunk 2]
```diff
@@ -121,8 +126,40 @@ public void setValues(UserRepresentation user) {
         setLastName(user.getLastName());
         if (user.isEnabled() != null) setEnabled(user.isEnabled());
         if (user.isEmailVerified() != null) setEmailVerified(user.isEmailVerified());
+        if (user.getGroups() != null && user.getGroups().length > 0) setGroups(new HashSet<String>(user.getGroups()));
         if (user.getRequiredActions() != null) setRequiredActions(new HashSet<>(user.getRequiredActions()));
     }
 
     // TODO Contact Information section
+
+    public class GroupSelect extends MultipleStringSelect2 {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/users/UserAttributesForm.java:135]: I'm getting some compilation errors here. The tests need to be passing with Java 8 (which is used by our CI).

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/groups/GroupTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    @Page
+    private CreateGroup createGroupPage;
+
+    @Before
+    public void beforeCreateGroupTest() {
+        groupsPage.navigateTo();
+    }
+
+    @Test
+    public void createAGroup() {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/groups/GroupTest.java:43]: This test seem to be failing. Needs to be passing with FF and Chrome.

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/users/UserAttributesTest.java [hunk 4]
```diff
@@ -99,4 +112,31 @@ public void disabledUser() {
         // TODO try to log in
     }
 
+    @Test
+    public void createUserWithGroups() {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/users/UserAttributesTest.java:116]: This test seem to be failing. Needs to be passing with FF and Chrome.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@clementcur Thank you for your contribution! I've added a few comments.

File: themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html [hunk 5]
```diff
@@ -97,6 +97,42 @@
...
+                <label class="col-md-2 control-label" for="selected-groups">{{:: 'groups-joining' | translate}}</label>
+                <div class="col-sm-6">
+                    <table class="table table-striped table-bordered" style="margin-top: 0px" id="selected-groups">
+                        <thead>
+                        <tr>
+                            <th class="kc-table-actions" colspan="2">
+                                <div class="form-inline col-md-12" style="width: 107%">
+                                    <div class="form-group" style="width: 100%">
+                                        <div class="input-group" style="width: 100%">
+                                            <input type="hidden" ui-select2="groupsUiSelect" id="groups" data-ng-change="selectGroup(selectedGroup);" data-ng-model="selectedGroup" data-placeholder="{{:: 'groups-joining-select.placeholder' | translate}}..."/>
```
[the remark this answers is not in the document]
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: I've copy-pasted a component used in Permissions UI:
image

IMO, a tabular view better highlights groups hierarchy, compared to a "tags input".

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js [hunk 6]
```diff
@@ -513,6 +515,39 @@ module.controller('UserDetailCtrl', function($scope, realm, user, BruteForceUser
         }
     }
 
+    function pushSelectedGroupsToUser() {
+        var groups = $scope.user.groups;
+        if ($scope.selectedGroups) {
```
[the remark this answers is not in the document]
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js:520]: This UI is driven by the REST Admin API. If we want a special behavior, we'll need to change that in the API itself. IMO, this is a different topic.

File: themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: Yes, you're correct. But there it works a bit differently. Once you've clicked a policy in the search bar, it appears in the table below and the search bar is clears. There are no duplicit "remove" buttons.

[the remark this answers is not in the document]
Comment (ssilvert) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: @vmuzikar The `x` means "clear the field".  That's different from remove and it's the way PatternFly does it.

The Required Actions solution uses chip groups, which are not appropriate here because chip groups get very hard to manage when you have more than a few items.  See the Login Events Settings for a really horrific example of this.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js [hunk 7]
```diff
@@ -527,6 +527,27 @@ module.controller('UserDetailCtrl', function($scope, realm, user, BruteForceUser
...
+    	}
+    	return;
+    }
+
+    function flattenGroups(groups) {
+        var flattenedGroups = [];
+        if (!groups || groups.length === 0) return groups;
+        for (var i=0; i < groups.length; i++) {
+            flattenedGroups.push(groups[i]);
+            bfs(groups[i], "subGroups", flattenedGroups);
```
Comment (ssilvert) [inline themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js:545]: I don't think you need to pass in "subGroups".  I don't think that value can ever change.

File: themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (ssilvert) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: The group selector is still not quite working right.  For instance, here there is no "s" in `/group1`, but it shows that group anyway.  Looks like it's because it has a subgroup with an "s" in it.
group-select

[the remark this answers is not in the document]
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: > @vmuzikar The `x` means "clear the field". That's different from remove and it's the way PatternFly does it.

@ssilvert Yes, I know what the "X" does but that might not be the case with the user. It might give the impression that the "X" removes the selection and therefore user might be unaware that this is a multi-select box.

> The Required Actions solution uses chip groups, which are not appropriate here because chip groups get very hard to manage when you have more than a few items. See the Login Events Settings for a really horrific example of this.

I assumed that no more that – let's say – 3 groups would be added here. It's still a manual process and IMHO doesn't make much sense for the admin to add tens of groups for just one user. But ok, the chip groups might not be the best solution.

I believe we should at least unify it with the Permissions UI – clear the search box after selection.

[the remark this answers is not in the document]
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: > The group selector is still not quite working right. For instance, here there is no "s" in `/group1`, but it shows that group anyway. Looks like it's because it has a subgroup with an "s" in it.

This is what our REST API returns so I'd say it is out of scope of this PR. It's probably implemented like that because of the "tree" UI the Groups section uses. When you search for a group, you need to see all its parents to be able to properly display the tree.

Maybe we could do some filtering at the UI side but I'm not sure that's the correct approach.

[the remark this answers is not in the document]
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: It's correctly aligned on my side and seems OK on ssilvert's screenshot too. Maybe check if you're not zoomed in.
The html code is strictly copy-pasted from the Permissions UI.

[the remark this answers is not in the document]
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html:110]: I'm not zoomed in. :) It's ok in Firefox, misaligned in Chrome and Safari.

[the remark this answers is not in the document]
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html]: I've fixed the alignment in the latest commit.

[the remark this answers is not in the document]
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/partials/user-detail.html]: I've flatten and filtered out searched groups in latest commit. Also, I made sure the input is cleared after selection, so there's no confusion.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js:521]: Since every node of the groups tree is now listed in search results, it's possible to add both parent and child groups.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js (hunk 7 above)
Comment (clementcur) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/users.js]: Fixed in my latest commit.

File: testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/users/UserAttributesForm.java (hunk 2 above)
Comment (clementcur) [reply testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/users/UserAttributesForm.java:141]: fixed in latest commit

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/groups/GroupTest.java (hunk 3 above)
Comment (clementcur) [reply testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/groups/GroupTest.java:43]: fixed in latest commit

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/users/UserAttributesTest.java (hunk 4 above)
Comment (clementcur) [reply testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/users/UserAttributesTest.java:117]: fixed in latest commit

**Comment by clementcur** [comment]
I've addressed all comments in the latest 3 commits, let me know what you guys think @ssilvert, @vmuzikar & @stianst.

**Comment by clementcur** [comment]
ping @ssilvert, @vmuzikar & @stianst

**Comment by ssilvert** [comment]
I'm happy with this from a UI standpoint.  Just running the admin console tests as a last check.  Build number 206.

**Comment by clementcur** [comment]
@ssilvert I cannot find the build #206 in Travis CI, let me know if everything's OK.

**Comment by ssilvert** [comment]
> @ssilvert I cannot find the build #206 in Travis CI, let me know if everything's OK.

I think you won't be able to see it because it is behind our firewall.  For the record, I had to restart, so the build was number 207.  

I compared to a test run against master and found that this PR introduces test failures in Firefox.  There should be only one failure, which is org.keycloak.testsuite.console.clients.ClientMappersOIDCTest.testUpdateTokenClaimName.  But with Firefox, these additional tests are now failing:


Test Name | Duration | Age
-- | -- | --
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testRegexPatternsPolicy | 6.4 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testUpperCasePolicy | 6 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testDigitsPolicy | 5.7 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testNotUsernamePolicy | 6.1 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testLengthPolicy | 6.1 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testSpecialCharsPolicy | 6.3 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testLowerCasePolicy | 6.1 sec | 1
org.keycloak.testsuite.console.authentication.PasswordPolicyTest.testPasswordHistoryPolicy | 5.9 sec | 1
org.keycloak.testsuite.console.federation.LdapUserFederationTest.checkVendors | 4.4 sec | 1
org.keycloak.testsuite.console.users.RequiredUserActionsTest.updateProfile | 9.4 sec | 1
org.keycloak.testsuite.console.users.RequiredUserActionsTest.updatePassword | 4.2 sec | 1
org.keycloak.testsuite.console.users.UserAttributesTest.createUserWithGroups | 7.3 sec | 1
org.keycloak.testsuite.console.federation.KerberosUserFederationTest.configureKerberosProvider | 10 sec | 3
org.keycloak.testsuite.console.federation.KerberosUserFederationTest.invalidSettingsTest | 4.8 sec | 3
org.keycloak.testsuite.console.federation.LdapUserFederationTest.testConnection | 9.6 sec | 3
org.keycloak.testsuite.console.federation.LdapUserFederationTest.configureConnectionPooling | 7.3 sec | 3
org.keycloak.testsuite.console.federation.LdapUserFederationTest.configureAdProvider | 7.5 sec | 3
org.keycloak.testsuite.console.federation.LdapUserFederationTest.invalidSettingsTest | 7.3 sec | 3
org.keycloak.testsuite.console.federation.LdapUserFederationTest.configureRhdsProvider | 7.4 sec | 3
org.keycloak.testsuite.console.federation.UserFederationTest.availableProviders | 8.4 sec | 3
org.keycloak.testsuite.console.federation.UserFederationTest.addRemoveFederation | 6.1 sec | 3
org.keycloak.testsuite.console.realm.SecurityDefensesTest.userUnlockTest | 12 sec | 3
org.keycloak.testsuite.console.roles.DefaultRolesTest.defaultRoleAssignedToNewUser | 12 sec | 3
org.keycloak.testsuite.console.users.RequiredUserActionsTest.termsAndConditions | 5.5 sec | 3
org.keycloak.testsuite.console.users.UserAttributesTest.disabledUser | 9.1 sec | 3
org.keycloak.testsuite.console.users.UserAttributesTest.createUserEmailAsUserName | 4.6 sec | 3
org.keycloak.testsuite.console.users.UserAttributesTest.invalidEmail | 4.4 sec | 3
org.keycloak.testsuite.console.users.UserAttributesTest.existingUser | 4.2 sec | 3
org.keycloak.testsuite.console.users.UserAttributesTest.noUsername | 3.6 sec | 3

I'm going to run this again to double-check.  I don't see why it would cause those failures.

**Comment by ssilvert** [comment]
Kicked off build 209.

**Review by ssilvert — DISMISSED** [review]

**Comment by ssilvert** [comment]
> Kicked off build 209.

Same odd result running against master.  So I'm convinced that this PR is not to blame.

**Comment by ssilvert** [comment]
@stianst I think your outdated change request is blocking a merge. 

I'm happy with this now as long as @vmuzikar agrees.

**Review by vmuzikar — COMMENTED** [review]
@clementcur Thanks for the update! Changes look good to me now.
Since the tests ran by @ssilvert failed due to some CI issues, I'd like to re-run them. However, we need to wait till the issues are fixed. Please wait with the merge a bit more. Thank you.

**Comment by clementcur** [comment]
thanks @vmuzikar, any news ?

**Comment by vmuzikar** [comment]
@clementcur I'm sorry for the late reply. I ran the whole test pipeline.
The good news is it looks there are no regressions.
The bad news is it looks like the Admin Console UI tests are broken in FF (Chrome is fine):
```
[ERROR] Failures: 
[ERROR]   LdapUserFederationTest.checkVendors:175 Expected vendors count: 5; actual count: 0
[ERROR]   UserFederationTest.availableProviders:30
[ERROR] Errors: 
[ERROR]   PasswordPolicyTest.testDigitsPolicy:91 » NoSuchElement Unable to locate elemen...
[ERROR]   PasswordPolicyTest.testLengthPolicy:77 » NoSuchElement Unable to locate elemen...
[ERROR]   PasswordPolicyTest.testLowerCasePolicy:105 » NoSuchElement Unable to locate el...
[ERROR]   PasswordPolicyTest.testNotUsernamePolicy:147 » NoSuchElement Unable to locate ...
[ERROR]   PasswordPolicyTest.testPasswordHistoryPolicy:175 » NoSuchElement Unable to loc...
[ERROR]   PasswordPolicyTest.testRegexPatternsPolicy:161 » NoSuchElement Unable to locat...
[ERROR]   PasswordPolicyTest.testSpecialCharsPolicy:133 » NoSuchElement Unable to locate...
[ERROR]   PasswordPolicyTest.testUpperCasePolicy:119 » NoSuchElement Unable to locate el...
[ERROR]   ClientMappersOIDCTest.testUpdateTokenClaimName:392 » NoSuchElement Unable to l...
[ERROR]   KerberosUserFederationTest.configureKerberosProvider:32 » NoSuchElement Unable...
[ERROR]   KerberosUserFederationTest.invalidSettingsTest:50 » NoSuchElement Unable to lo...
[ERROR]   LdapUserFederationTest.configureAdProvider:39 » NoSuchElement Cannot locate el...
[ERROR]   LdapUserFederationTest.configureConnectionPooling:184 » NoSuchElement Cannot l...
[ERROR]   LdapUserFederationTest.configureRhdsProvider:70 » NoSuchElement Cannot locate ...
[ERROR]   LdapUserFederationTest.invalidSettingsTest:92 » NoSuchElement Cannot locate el...
[ERROR]   LdapUserFederationTest.testConnection:137 » NoSuchElement Cannot locate option...
[ERROR]   UserFederationTest.addRemoveFederation:38 » NoSuchElement Cannot locate elemen...
[ERROR]   SecurityDefensesTest.userUnlockTest:163 » NoSuchElement Unable to locate eleme...
[ERROR]   DefaultRolesTest.defaultRoleAssignedToNewUser:57 » NoSuchElement Unable to loc...
[ERROR]   RequiredUserActionsTest.termsAndConditions:134 » NoSuchElement Unable to locat...
[ERROR]   RequiredUserActionsTest.updatePassword:76 » NoSuchElement Unable to locate ele...
[ERROR]   RequiredUserActionsTest.updateProfile:100 » NoSuchElement Unable to locate ele...
[ERROR]   UserAttributesTest.createUserEmailAsUserName:80->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.createUserWithGroups:139->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.disabledUser:111->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.existingUser:94->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.invalidEmail:66->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.noUsername:86->AbstractUserTest.createUser:36 » NoSuchElement
```
I re-ran the tests twice to be sure, both times with the same result. It's not happening in master, there's "only" 1 test failing both with FF and Chrome (`ClientMappersOIDCTest.testUpdateTokenClaimName:392`).

I'm not sure what's causing it. It's possible it won't be reproducible locally – our CI is a bit slower so some timing-sensitive errors manifest only there.

**Comment by vmuzikar** [comment]
@clementcur Maybe a rebase could help here...

**Comment by clementcur** [comment]
I just rebased, let me know how it goes @vmuzikar

**Comment by clementcur** [comment]
ping @vmuzikar

**Comment by vmuzikar** [comment]
@clementcur Sorry, haven't had yet time to look at the latest changes. Will do the next week. Thanks.

**Comment by clementcur** [comment]
hey @vmuzikar, any chance you can have a look?

**Comment by vmuzikar** [comment]
@clementcur I'm sorry for a very late response. I was busy with other tasks.

Unfortunately, Admin Console UI tests are still broken in FF (Chrome is not affected):
```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   LdapUserFederationTest.checkVendors:175 Expected vendors count: 5; actual count: 0
[ERROR]   UserFederationTest.availableProviders:30
[ERROR] Errors: 
[ERROR]   PasswordPolicyTest.testDigitsPolicy:91 » NoSuchElement Unable to locate elemen...
[ERROR]   PasswordPolicyTest.testLengthPolicy:77 » NoSuchElement Unable to locate elemen...
[ERROR]   PasswordPolicyTest.testLowerCasePolicy:105 » NoSuchElement Unable to locate el...
[ERROR]   PasswordPolicyTest.testNotUsernamePolicy:147 » NoSuchElement Unable to locate ...
[ERROR]   PasswordPolicyTest.testPasswordHistoryPolicy:175 » NoSuchElement Unable to loc...
[ERROR]   PasswordPolicyTest.testRegexPatternsPolicy:161 » NoSuchElement Unable to locat...
[ERROR]   PasswordPolicyTest.testSpecialCharsPolicy:133 » NoSuchElement Unable to locate...
[ERROR]   PasswordPolicyTest.testUpperCasePolicy:119 » NoSuchElement Unable to locate el...
[ERROR]   ClientMappersOIDCTest.testUpdateTokenClaimName:392 » NoSuchElement Unable to l...
[ERROR]   KerberosUserFederationTest.configureKerberosProvider:32 » NoSuchElement Unable...
[ERROR]   KerberosUserFederationTest.invalidSettingsTest:50 » NoSuchElement Unable to lo...
[ERROR]   LdapUserFederationTest.configureAdProvider:39 » NoSuchElement Cannot locate el...
[ERROR]   LdapUserFederationTest.configureConnectionPooling:184 » NoSuchElement Cannot l...
[ERROR]   LdapUserFederationTest.configureRhdsProvider:70 » NoSuchElement Cannot locate ...
[ERROR]   LdapUserFederationTest.invalidSettingsTest:92 » NoSuchElement Cannot locate el...
[ERROR]   LdapUserFederationTest.testConnection:137 » NoSuchElement Cannot locate option...
[ERROR]   UserFederationTest.addRemoveFederation:38 » NoSuchElement Cannot locate elemen...
[ERROR]   SecurityDefensesTest.userUnlockTest:163 » NoSuchElement Unable to locate eleme...
[ERROR]   DefaultRolesTest.defaultRoleAssignedToNewUser:57 » NoSuchElement Unable to loc...
[ERROR]   RequiredUserActionsTest.termsAndConditions:134 » NoSuchElement Unable to locat...
[ERROR]   RequiredUserActionsTest.updatePassword:76 » NoSuchElement Unable to locate ele...
[ERROR]   RequiredUserActionsTest.updateProfile:100 » NoSuchElement Unable to locate ele...
[ERROR]   UserAttributesTest.createUserEmailAsUserName:80->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.createUserWithGroups:139->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.disabledUser:111->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.existingUser:94->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.invalidEmail:66->AbstractUserTest.createUser:36 » NoSuchElement
[ERROR]   UserAttributesTest.noUsername:86->AbstractUserTest.createUser:36 » NoSuchElement
```
It's happening only in this PR, so it must be caused by it after all.

This must be a bit frustrating for you. Let me know if you were able to reproduce it locally and find the culprit. If not, maybe I would be able to help somehow. Thank you for your endurance! :)

**Comment by clementcur** [comment]
Hi @vmuzikar, I just rebased, reran the whole UI test suite using Firefox user-agent and got a success build...

Can you tell me if you still see issues on your CI?

```
[INFO]
[INFO] Results:
[INFO]
[WARNING] Tests run: 184, Failures: 0, Errors: 0, Skipped: 5
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:31 h
[INFO] Finished at: 2020-12-18T18:19:19+01:00
[INFO] ------------------------------------------------------------------------
```

**Comment by clementcur** [comment]
@vmuzikar, all checks have passed ! yay ! Let me know if it's all good on your side.

**Comment by clementcur** [comment]
ping @vmuzikar

**Review by vmuzikar — APPROVED** [review]
@clementcur Thank you for the update and sorry for delayed response during the Christmas break.
Unfortunately, Firefox is currently broken in our CI so I couldn't re-run the test there. I ran it at least locally and it seems ok. I don't want to block this PR anymore and let it rot even more, so I'm gonna approve.

Thank you for your contribution!

**Comment by clementcur** [comment]
thanx @vmuzikar ! however, merging is blocked by an already addressed requested change, @stianst can you remove the requested change ?

**Comment by clementcur** [comment]
@stianst, merging is blocked, can you remove the requested change ?

**Comment by clementcur** [comment]
@vmuzikar, any chance you can ask @stianst to remove the requested change ?

**Comment by stianst** [comment]
@ssilvert @pedroigor can you guys review this one please?

**Comment by ssilvert** [comment]
> @ssilvert @pedroigor can you guys review this one please?

Tried it out again and it looks good.  Just need to make sure the tests run ok.

**Comment by clementcur** [comment]
@stianst, @ssilvert, @pedroigor all good?

**Comment by pedroigor** [comment]
@clementcur Tests are running fine for me. However, if we could check if the user was indeed added to the group after creating the user, better.

Otherwise, LGTM.

@ssilvert I guess you are waiting for the test results from our testing pipeline?

**Review by pedroigor — APPROVED** [review]

**Comment by ssilvert** [comment]
> @clementcur Tests are running fine for me. However, if we could check if the user was indeed added to the group after creating the user, better.
> 
> Otherwise, LGTM.
> 
> @ssilvert I guess you are waiting for the test results from our testing pipeline?

Yes.  The way we used to run the tests is broken.  I'm working with QE to get it done.

**Comment by clementcur** [comment]
@ssilvert, any update ?

**Comment by ssilvert** [comment]
> @ssilvert, any update ?

I finally ended up running it locally and the tests look good.  Approved.

**Review by ssilvert — APPROVED** [review]

**Comment by ssilvert** [comment]
@stianst,

Looks like everyone is happy with this now.

**Comment by clementcur** [comment]
@stianst can you approve and remove the change requested if everything looks fine ?
